### PR TITLE
Parse filter

### PR DESCRIPTION
**Pull request by psieg**
second attempt. Turns out the travis environment doesn't like Sets but prefers complaining about dots....
#1 #4

before merge
- [x] confirm solver capabilities with tim
- [x] figure out which of the layer code is needed
- [ ] _educational:_ and why there is this layer-testing in visitGetSlot
- [x] add some tests
- [x] detect finite domains
- [x] detect types through method calls
  - [x] confirm this actually works with the temperature example

### CODE REVIEW

**Comment by j4yk** [comment]
> Turns out the travis environment doesn't like Sets but prefers complaining about dots....

:+1:

File: babelsberg/constraintinterpreter.js [hunk 1]
```diff
@@ -375,9 +390,20 @@ Object.subclass('Babelsberg', {
...
+                            ' because it does not support ' + type + ' variables');
+                    }
+                    return false;
+                }
+            }
+
             result.push(solver);
         });
 
+        delete seenTypes;
```
Comment (j4yk) [inline babelsberg/constraintinterpreter.js]: Shouldn't this read `delete bbb.seenTypes`?
Not sure what V8's hidden classes make out of this, maybe we should simply set it to null or to empty again.

File: babelsberg/tests.js [hunk 2]
```diff
@@ -2061,6 +2061,83 @@ TestCase.subclass('users.timfelgentreff.babelsberg.tests.AutomaticSolverSelectio
         this.assert(man.shirt !== man.pants, "shirt and pants must not have the same color");
         this.assert(man.pants === "black" || man.pants === "blue" || man.pants === "white", "pants should be 'black', 'blue' or 'white'");
     },
+	testFilteringByPriority: function () {
```
Comment (j4yk) [inline babelsberg/tests.js]: one tab here, and more below on the test... declarations
(the linter does not check the tests as you can see)

File: babelsberg/tests.js [hunk 3]
```diff
@@ -2061,6 +2061,83 @@ TestCase.subclass('users.timfelgentreff.babelsberg.tests.AutomaticSolverSelectio
...
+        testCase = this;
```
Comment (j4yk) [inline babelsberg/tests.js]: `var testCase = ...` or testCase will be global

File: babelsberg/tests.js (hunk 2 above)
Comment (psieg) [inline babelsberg/tests.js]: tabs vs. spaces is horribly messed up in these files. I'd rather do it in another PR to keep the diff readable (yes I know about the spaces flag)

**Comment by psieg** [comment]
@timfel can you confirm our choice for the solver capabilities?

**Comment by timfel** [comment]
supportsFiniteDomains would be good, too. It's only used with csp and backtalk, so far

**Comment by timfel** [comment]
Other than that, looks good

**Comment by j4yk** [comment]
finite domain definitions manifest as "in" in the constraint expression

`pt.x.is in [1, 2, 7, 9]`

**Comment by psieg** [comment]
@j4yk are you aware the backtalk example with automatic selection is solved by DeltaBlue and CSP?

**Comment by j4yk** [comment]
@psieg Yes I am. We have already talked about that: the test is named backtalkExample because it stems from the backtalk paper, but I copied the test function from the test cases for CSP ;-) Also I think it is not important which solver solves this test, but it should only work, because that is the goal of the automatic selection.

**Comment by psieg** [comment]
I tried using the Thermometer example with this branch and ended up with the cassowary problem:
- We check types and find strings (the text values of the labels)
- Hardcoded "correct" choice and seemingly only defaultSolver capable of this constraint is cassowary
- According to discussion with tim today, cassowary does not support strings except through js-builtin coercing and should not be listed as "supports strings" - however that is exactly what happens in this example
- Result: With our algorithm designed to help you find the correct solver, we take the only working solver out of your hands because we clam it does not support strings.

Also:
- We detect the use of "Methods" through existence of `opts.methods`. In the most common cases however, the propagation functions can be implicitly found by Deltablue (example: second half of thermometer example), thus there is no opts.methods and we do not rule out e.g. Cassowary

**Comment by j4yk** [comment]
Then let's pretend Cassowary did support strings and add a comment "not really but...".

More advanced: can we detect syntactically (in the InterpreterVisitor) that automatic string conversion will take place?

I do not see why the "Also" part should be a problem. Either Cassowary can indeed fulfill the methodless constraint as well as DeltaBlue or better and can be the right choice, or it will not work and DeltaBlue is chosen again.

**Comment by psieg** [comment]
About the also: It will not (afaik), so this _could_ be prevented by a better filtering. The purpose if this feature is to (a) have the user learn which solver does which and (b) reduce computation time spent trying solvers that can't do it
Unfortunately I don't have the feeling this branch helps too much in that regard
